### PR TITLE
add storage-initializer uid handling for OpenShift with istio-cni

### PR DESCRIPTION
**Pull request by ReToCode**
**What this PR does / why we need it**:
- This PR adds custom code to make KServe run on OpenShift without the need for `anyuid` SCC.

More context:
OpenShift uses istio-cni which causes an issue with init-containers when calling external services
like S3 or similar. Setting the `uid` for the `storage-initializer` to the same `uid` as the
`uid` of the `istio-proxy` resolves the issue. In OpenShift the `istio-proxy` always gets assigned
the first `uid` from the namespaces `uid` range + 1 (The range is defined in an annotation on the namespace).

**Type of changes**
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:
Tested on OpenShift according to documentation in https://github.com/ReToCode/knative-kserve and https://github.com/opendatahub-io/odh-manifests/pull/838.

**Logs**
Installation setup and documentation [here](https://github.com/ReToCode/knative-kserve/tree/main).

**Checklist**:

- [X] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [X] Has code been commented, particularly in hard-to-understand areas?
- [X] Have you made corresponding changes to the documentation?

**Release note**:
```release-note
The `storage-initializer` container will now run with the same `uid` as the `istio-proxy` which resolves an issue when istio-cni is used.
```

/hold WIP - making tests pass.

### CODE REVIEW

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (danielezonca) [inline pkg/webhook/admission/pod/storage_initializer_injector.go:345]: Who is it adding this annotation? Istio? Can you add this context to the error message?

File: pkg/webhook/admission/pod/mutator.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/mutator.go:66]: I just followed the pattern in the [existing file](https://github.com/opendatahub-io/kserve/pull/18/files/faf4dfd3e74f050b4e50df95861e95682abe9e3a#diff-e6409d7113f94986ff06bbce8910aae0f2b255e0e635e48452c146870fb8595dR56). I'm fine both ways.

File: pkg/webhook/admission/pod/mutator.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/mutator.go:68]: Same as in https://github.com/opendatahub-io/kserve/pull/18/files/faf4dfd3e74f050b4e50df95861e95682abe9e3a#r1259364584

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:343]: Hm not sure if it is better to fail or not (if the annotation is not there, something went pretty wrong on OCP). As the user in OCP probably will never know about the issue and how to resolve it (needing to know to add 1 for example), do we expect that he even could bring the correct user? I'm currently favouring erroring I think.

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:345]: That is an annotation OpenShift adds to every namespace. But yes, I can add it to the message.

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (rhuss) [inline pkg/webhook/admission/pod/storage_initializer_injector.go:340]: Wouldn't it be better to move this functionality in a dedicated initializer instead of putting it into ta "storage initializer" ? Semantically I don't think its a good fit here to mangle with the security context. It also should be easier to add/remove an initializer at an higher level instead of modifying an initializer (and from the outside its not clear that a storage related initializer also manages the security context)

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (rhuss) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:343]: I don't think that this mechanism with "+1" would work as Istio itself would not find that range, right ? I'm not sure what Istio/Service Mesh would do as a fallback when the range is not available as an annotation. I would then probably do the same here, too.

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (rhuss) [inline pkg/webhook/admission/pod/storage_initializer_injector.go:353]:
```suggestion
		return fmt.Errorf("could not find OpenShift internal annotation %s for calculating the process UID (minRange + 1) on namespace: %s", OpenShiftUidRangeAnnotationKey, targetNs.Name)
	}
	uidStrParts := strings.Split(uidStr, "/")
	if uid, err := strconv.ParseInt(uidStrParts[0], 10, 64); err == nil {
		// Set the uid to the first uid in the namespaces range + 1
		uid++
		initContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
	} else {
		return fmt.Errorf("could not parse value %s as an int64 value in annotation %s on namespace: %s", uidStr, OpenShiftUidRangeAnnotationKey, targetNs.Name)
```

I'm always like adding as much context in error messages as possible. Always remember that this is probably read by someone who has no clue about the algo.

File: pkg/webhook/admission/pod/mutator.go
Comment (rhuss) [inline pkg/webhook/admission/pod/mutator.go:126]: Instead of changing the signature of every injector by a namespace argument that mostly is ignored except for one injector, I would instead use a closure for the one injector that needs a namespace and leave the signature of the others unchanged, to avoid unnecessary churn. Especially that you already know that namespace at this point

E.g. 

```
mutators := []func(pod *v1.Pod) error {
                InjectGKEAcceleratorSelector,
		func(pod *v1.Pod) error { 
                     return storageInitializer.InjectStorageInitializer(pod, targetNs)
                },
		agentInjector.InjectAgent,
		metricsAggregator.InjectMetricsAggregator,
	}
```

(but as said elsewhere, I would use a dedicated initalizer for this functionality instead of mangling it into the storage initializer)

Comment (rhuss) [reply pkg/webhook/admission/pod/mutator.go:126]: And, tbh, I don't know why using this complex way of calling for method is used anyways ;-)

Just calling one after each other inline would be much more straight forward in this case. If you would move those list of mutators around method boundaries, then it might make sense, but here it's just overengineered.

I suggest to just use:

```
err := injectGKEAcceleratorSelector(pod);
if err != nil {
   return err
}
err := injectGKEAcceleratorSelector(pod);
if err != nil {
   return err
}
err = storageInitializer.InjectStorageInitializer(pod, targetNs);
if err != nil {
   return err
}
err = agentInjector.InjectAgent(pod);
if err != nil {
   return err
}	
return metricsAggregator.InjectMetricsAggregator;
```

Comment (rhuss) [reply pkg/webhook/admission/pod/mutator.go:126]: I would also question that a method named "injectGKEAccelarator" is a good fit or such a high-level package (while all other injectors are in separate packages, this should be also in a separate package, too)

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:343]: Yeah, I think it should be fine to assume it is there or error out.

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:340]: The change is only relevant for the `storage-initializer` and these are the bits of code that create the `init-container` spec for that. I agree with you semantically, but having it somewhere else would mean that we have to know how to filter the `storage-initializer` and modifying its security context there. I find it more clear here.

Comment (rhuss) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:340]: Just saw that this is coming from upstream and shouldn't be changed here but in upstream kserve. So forget this comment for now.

File: pkg/webhook/admission/pod/mutator.go
Comment (rhuss) [reply pkg/webhook/admission/pod/mutator.go:126]: Just realized that this "abstraction" is coming from upstream, so shouldn't be discussed here. So, I conclude for a minimal change I suggest the solution with the closure as this does not require changing the signatures of the other initializers.

**Review by danielezonca — COMMENTED** [review]
/lgtm

I would like to have @Jooho or @israel-hdez approval too before merging

**Comment by danielezonca** [comment]
@ReToCode 
Is the PR still WIP or it is just the title to be updated?

@Jooho @israel-hdez Please review when you have time :)

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (israel-hdez) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:343]: I think I'm more in agreement with @danielezonca here.

I think the code should honor the `IstioSidecarUIDAnnotationKey` if present. Otherwise, apply OpenShift logic. If both the OpenShift annotations and `IstioSidecarUIDAnnotationKey` aren't there, continue without doing any change and without error .

That logic ^ would make the code compatible with both upstream Istio-cni (which will want the `1337` annotation) and Maistra/OSSM (which will use OpenShift annotations). And the code would also remain compatible with an upstream non-cni-Istio installation (which presumably works well without UID hacking), in case community wants to try that.

Comment (israel-hdez) [reply pkg/webhook/admission/pod/storage_initializer_injector.go:343]: ...I'm also thinking that such logic ^ (despite being a workaround) is perhaps more feasible to push to upstream, as it is less invasive.

**Review by israel-hdez — COMMENTED** [review]
I dropped a comment for your consideration: https://github.com/opendatahub-io/kserve/pull/18#discussion_r1260127612

File: pkg/webhook/admission/pod/storage_initializer_injector.go
Comment (ReToCode) [reply pkg/webhook/admission/pod/storage_initializer_injector.go]: I like your proposal and explanation. I think that makes sense. I updated the PR accordingly, please re-review.

**Comment by ReToCode** [comment]
> Is the PR still WIP or it is just the title to be updated?

I'm still testing it with quay.io/rlehmann/kserve-controller:fix-anyuid. I'm targeting it to be ready today.

**Comment by ReToCode** [comment]
I'm done with testing. Looks good. Please also see https://redhat-internal.slack.com/archives/C05742W6F7T/p1689149577193889 for a demo link.

**Comment by skonto** [comment]
Do we still have an option if the user actually wants to have anyuid in their namespace because they are running something privileged in there?

**Comment by ReToCode** [comment]
> If the user actually wants to have anyuid in their namespace because they are running something privileged in there they could still use IstioSidecarUIDAnnotationKey according to the PR (storage init container uid should be set to 1337 in that case). We could even automate that too right (although a bit more complex eg. detect sa privileges etc)?

Hm not sure if we even would want that, but you can run your user-container as privileged, the istio-container would still be 1337 or the OCP one (we just set the storage-initializer to the same id as istio). I doubt that there is a case for running the storage-initializer privileged. Or what case are you thinking of?

**Comment by skonto** [comment]
> we just set the storage-initializer to the same id as istio)

It seems to me that we only set it to project range uid + 1. If user sets anyuid istio side car will run with 1337 but storage-initializer will run with a project range id no?

**Comment by skonto** [comment]
> If you use Maistra/OSSM, AFAIK, the Istio sidecars will always use projectRangeUid + 1 regardless of the anyuid setting.

Ok if so makes sense then.

**Review by israel-hdez — APPROVED** [review]
Looks good.

**Comment by israel-hdez** [comment]
> The setup is done using Reto's repo and its Maistra/SM on OCP 4.12. As you see the istio sidecar uses 1337 in the second case.

That's quite interesting. Good find! So, I was wrong.

Digging on Maistra's code, for reference this is how they implement it: https://github.com/maistra/istio/blob/98f70f7641d14db1351c4b8f5ae1d8f19fc79a92/pkg/kube/inject/webhook.go#L922-L959. Looks like uid `1337` is a fallback.

> Probably some form of documentation would help here.

I'm quite in agreement with JooHo comment and this is supposed to be a workaround. Perhaps, if we see that this workaround won't go away, we will need to improve it. So, I'm :+1:  to cover it with docs for the time being.

I'll create a ticket.

**Pull request merged**
